src/uucore/src/lib/features/checksum.rs

```diff
@@ -2,26 +2,967 @@
+const SINGLE_SPACE_REGEX: &str = r"^(?P<checksum>[a-fA-F0-9]+)\s(?P<filename>\*?.*)$";
+
+fn get_filename_for_output(filename: &OsStr, input_is_stdin: bool) -> String {
+    if input_is_stdin {
+        "standard input"
+    } else {
+        filename.to_str().unwrap()
+    }
+    .maybe_quote()
+    .to_string()
+}
+
+/// Determines the appropriate regular expression to use based on the provided lines.
+fn determine_regex(
+    filename: &OsStr,
+    input_is_stdin: bool,
+    lines: &[String],
+) -> UResult<(Regex, bool)> {
+    let algo_based_regex = Regex::new(ALGO_BASED_REGEX).unwrap();
+    let double_space_regex = Regex::new(DOUBLE_SPACE_REGEX).unwrap();
+    let single_space_regex = Regex::new(SINGLE_SPACE_REGEX).unwrap();
+
+    for line in lines {
+        let line_trim = line.trim();
+        if algo_based_regex.is_match(line_trim) {
+            return Ok((algo_based_regex, true));
+        } else if double_space_regex.is_match(line_trim) {
+            return Ok((double_space_regex, false));
+        } else if single_space_regex.is_match(line_trim) {
+            return Ok((single_space_regex, false));
+        }
+    }
+
+    Err(
+        ChecksumError::NoProperlyFormattedChecksumLinesFound(get_filename_for_output(
+            filename,
+            input_is_stdin,
+        ))
+        .into(),
+    )
+}
+
+/***
+ * Do the checksum validation (can be strict or not)
+*/
+#[allow(clippy::too_many_arguments)]
+#[allow(clippy::cognitive_complexity)]
+pub fn perform_checksum_validation<'a, I>(
+    files: I,
+    strict: bool,
+    status: bool,
+    warn: bool,
+    binary: bool,
+    ignore_missing: bool,
+    quiet: bool,
+    algo_name_input: Option<&str>,
+    length_input: Option<usize>,
+) -> UResult<()>
+where
+    I: Iterator<Item = &'a OsStr>,
+{
+    // if cksum has several input files, it will print the result for each file
+    for filename_input in files {
+        let mut bad_format = 0;
+        let mut failed_cksum = 0;
+        let mut failed_open_file = 0;
+        let mut correct_format = 0;
+        let mut properly_formatted = false;
+        let input_is_stdin = filename_input == OsStr::new("-");
+
+        let file: Box<dyn Read> = if input_is_stdin {
+            Box::new(stdin()) // Use stdin if "-" is specified
+        } else {
+            match File::open(filename_input) {
+                Ok(f) => Box::new(f),
+                Err(_) => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!(
+                            "{}: No such file or directory",
+                            filename_input.to_string_lossy()
+                        ),
+                    )
+                    .into());
+                }
+            }
+        };
+
+        let reader = BufReader::new(file);
+        let lines: Vec<String> = reader.lines().collect::<Result<_, _>>()?;
+        let (chosen_regex, algo_based_format) =
```

Comment:
I would rename `algo_based_format` to something like `is_algo_based_format` so that it's obvious it's a bool.